### PR TITLE
beginning of conversion of src files to ts, utilising interfaces

### PR DESCRIPTION
**Pull request by jmuzsik**
Hi, this is a more informed TS effort!

I added `super` in `setDragImage` test because my tests continually failed otherwise while ESNext was on.

Otherwise, lots of type checks to get rid of most of the warnings that existed in each file. Most files do not have warnings atm.

Though, this is going to bring down test coverage as there are quite a few lines added without tests. I can add tests as well as soon as I can to help get that coverage up!

Also, the `tsconfig` has the part with `src/interfaces` as, I am quite sure, that is necessary to keep reference to `.d.ts` files.

Thanks!

### CODE REVIEW

File: src/attribute.ts [hunk 1]
```diff
@@ -1,9 +1,9 @@
 /**
- * @param {Array|Element} element
+ * @param {Array|elem} element
```
Comment (lukasoppermann) [inline src/attribute.ts]: Why are you changing it to elem? Shouldn't you rather change it to `Array<HTMLElement>|HTMLElement`?

File: src/attribute.ts [hunk 2]
```diff
@@ -1,9 +1,9 @@
...
  * @param {string} attribute
  * @param {*} value
```
Comment (lukasoppermann) [inline src/attribute.ts]: this should be `string` instead of `*` right?

File: src/attribute.ts [hunk 3]
```diff
@@ -1,9 +1,9 @@
...
  */
-function addAttribute (element, attribute, value) {
+function addAttribute (element: Array<Element>|Element, attribute:string, value:string) {
```
Comment (lukasoppermann) [inline src/attribute.ts]: I think it should be `element: Array<HTMLElement>|HTMLElement`.

File: src/attribute.ts [hunk 4]
```diff
@@ -16,7 +16,7 @@ function addAttribute (element, attribute, value) {
  * @param {Array|Element} element
  * @param {string} attribute
  */
-function removeAttribute (element, attribute) {
+function removeAttribute (element: Array<Element>|Element, attribute:string) {
```
Comment (lukasoppermann) [inline src/attribute.ts]: Same as above: `Array<HTMLElement>|HTMLElement`

File: src/data.ts [hunk 5]
```diff
@@ -2,10 +2,11 @@
  * Get or set data on element
  * @param {Element} element
  * @param {string} key
- * @param {*} value
+ * @param {string} value
```
Comment (lukasoppermann) [inline src/data.ts]: This needs to be `any`

File: src/data.ts [hunk 6]
```diff
@@ -2,10 +2,11 @@
  * Get or set data on element
  * @param {Element} element
```
Comment (lukasoppermann) [inline src/data.ts]: We should be using `HTMLElement`

File: src/data.ts [hunk 7]
```diff
@@ -2,10 +2,11 @@
...
  * @param {string} key
- * @param {*} value
+ * @param {string} value
  * @return {*}
  */
-function addData (element: Element, key: string, value?: any) {
+
+function addData (element: Element, key: string, value?: string): Element|opts|string|void {
```
Comment (lukasoppermann) [inline src/data.ts]: Same here: `HTMLElement` and value must be `any`

File: src/elementHeight.ts [hunk 8]
```diff
@@ -3,7 +3,7 @@
  * Get height of an element including padding
  * @param {Element} element an dom element
  */
-export default element => {
+export default (element: Element) => {
```
Comment (lukasoppermann) [inline src/elementHeight.ts]: Again `HTMLElement` 😉

File: src/debounce.ts [hunk 9]
```diff
@@ -1,13 +1,13 @@
 /**
  * Creates and returns a new debounced version of the passed function which will postpone its execution until after wait milliseconds have elapsed
  * @param {func} Function to debounce
- * @param {wait} time to wait before calling function with latest arguments, 0 - no debounce
+ * @param {number} time to wait before calling function with latest arguments, 0 - no debounce
  * @returns {function} - debounced function
  */
 export default (func: Function, wait: number = 0): Function => {
-  let timeout
+  let timeout = wait
```
Comment (lukasoppermann) [inline src/debounce.ts]: this should not be wait

File: src/debounce.ts [hunk 10]
```diff
@@ -1,13 +1,13 @@
...
 
-  return (...args) => {
+  return (...args: Array<Element>) => {
```
Comment (lukasoppermann) [inline src/debounce.ts]: We can say what the arguments are, so we need to remove the `Array<Element>`

File: src/eventListener.ts [hunk 11]
```diff
@@ -19,7 +20,7 @@ function addEventListener (element, event, callback) {
  * @param {Array|Element} element
  * @param {Array|string} event
  */
-function removeEventListener (element, event) {
+function removeEventListener (element: Element, event: string) {
```
Comment (lukasoppermann) [inline src/eventListener.ts]: Element should be `Array<HTMLElement>|HTMLElement`

File: src/index.ts [hunk 12]
```diff
@@ -4,7 +4,7 @@
  * @param {Element} element
  * @returns {number}
  */
-export default (element: Element, elementList: HTMLCollection | NodeList | Array): number => {
+export default (element: Element, elementList: HTMLCollection | NodeList | Array<Element>): number => {
```
Comment (lukasoppermann) [inline src/index.ts]: `Array<HTMLElement>`

**Review by lukasoppermann — CHANGES_REQUESTED** [review]
Hey, thank you for all the work!
I will look into the other files as well. 

I would prefer to use `types` instead if `interfaces` as a folder name, like in this PR: https://github.com/lukasoppermann/html5sortable/pull/387/files#diff-86b55afca5305c6cfb991f9b13264803

From my research `types` is a more typical folder for this. Please let me know if you have concerns.

Also, I thought we figured it would be best to create individual files for each interface (similar to the one I have in this PR, although I don't have the `declare` part which I will look into 👍). Would you mind updating or is there a reason for this?

Can you please rebase, I did some changes, that will influence your code in `makePlaceholder.ts` & `insertHtmlElements.ts`.

File: src/filter.ts [hunk 13]
```diff
@@ -5,13 +5,14 @@
  * @param {String} selector
  * @returns {Array}
  */
-export default (nodes: NodeList|HTMLCollection|Array, selector: string): Array<Element> => {
+export default (nodes: NodeList|HTMLCollection|Array<Element>, selector: string): Array<Element> => {
```
Comment (lukasoppermann) [inline src/filter.ts]: This should be `Array<HTMLElement>`

File: src/filter.ts [hunk 14]
```diff
@@ -5,13 +5,14 @@
...
   if (!(nodes instanceof NodeList || nodes instanceof HTMLCollection || nodes instanceof Array)) {
     throw new Error('You must provide a nodeList/HTMLCollection/Array of elements to be filtered.')
   }
+  let elArray:Element[] = Array.from(nodes)
```
Comment (lukasoppermann) [inline src/filter.ts]: Does this actually solve type issues? It does not for me.

File: src/getHandles.ts [hunk 15]
```diff
@@ -4,7 +4,7 @@
  * @param {NodeList|HTMLCollection|Arra} items
  * @param {string} selector
  */
-export default (items: NodeList|HTMLCollection|Array, selector: string): Array<Element> => {
+export default (items: NodeList|HTMLCollection|Array<Element>, selector: string): Array<Element> => {
```
Comment (lukasoppermann) [inline src/getHandles.ts]: I updated this file, I think.

File: src/insertHtmlElements.ts [hunk 16]
```diff
@@ -5,7 +5,7 @@
  * @param {Element} newElement - element to be inserted
  * @param {String} position - insert before or after reference element
  */
-let insertNode = (referenceNode: any, newElement: any, position: String) => {
+let insertNode = (referenceNode: Element, newElement: Element, position: String) => {
```
Comment (lukasoppermann) [inline src/insertHtmlElements.ts]: I updated this file as well already, sorry.

File: src/offset.ts [hunk 17]
```diff
@@ -1,9 +1,4 @@
-export interface offsetObject { // eslint-disable-line no-undef
-  'left': number,
-  'right': number,
-  'top': number,
-  'bottom': number
-}
+// eslint-disable-line no-undef
```
Comment (lukasoppermann) [inline src/offset.ts]: This comment is not needed if you remove the interface.

File: src/setDragImage.ts [hunk 18]
```diff
@@ -1,5 +1,5 @@
 /* eslint-env browser */
-import {default as offset, offsetObject} from './offset' // eslint-disable-line no-unused-vars
+import {default as offset} from './offset' // eslint-disable-line no-unused-vars
```
Comment (lukasoppermann) [inline src/setDragImage.ts]: I think the comment `// eslint-disable-line no-unused-vars` can be removed if you remove `offsetObject `

**Review by lukasoppermann — CHANGES_REQUESTED** [review]

**Comment by jmuzsik** [comment]
Ok! Think this is a much better implementation then the previous one! Open to any more changes.

`Element` changed everywhere to `HTMLElement, maybe I shouldn't have done so in the `html5sortable` file, let me know. 

`Type` folder created. For the few interfaces currently that are used only in specific files I created unique files.

Also, you know that `.d.ts` files can automatically be created by running: `tsc --declaration --outDir ./src/{something}`
screen shot 2018-03-21 at 12 18 26 pm

I made a `main.d.ts` as extensions to objects that are externally created such as `HTMLElement` can be done as so. Also, since it is used in multiple files I thought it best to do the defining in one file. Perhaps it is my editor but if I define an interface one time it does not need to be imported. I do not think the change in the `tsconfig` makes any difference in this case actually as I tried with and without it.

Also, lemme know if you care for the type guard. I can delete it, and I do not think it ought to belong in the `types` folder if it stays put:[ type guards](https://stackoverflow.com/questions/32713452/user-defined-type-guards-typescript)

Thanks!

**Review by lukasoppermann — COMMENTED** [review]
Hey, you will definitely need to git rebase, my changes should not be in your PR. I added some more minor comments, will check further tomorrow. Looking good already. 👍

File: src/eventListener.ts [hunk 19]
```diff
@@ -16,10 +17,10 @@ function addEventListener (element, event, callback) {
   element.h5s.events[event] = callback
 }
 /**
- * @param {Array|Element} element
- * @param {Array|string} event
+ * @param {Array<HTMLElement>|HTMLElement} element
+ * @param {Array<HTMLElement>|string} event
  */
-function removeEventListener (element, event) {
+function removeEventListener (element: HTMLElement, event: string) {
```
Comment (lukasoppermann) [inline src/eventListener.ts:22]: This should also be `element: Array<HTMLElement>|HTMLElement` I think.

File: src/filter.ts [hunk 20]
```diff
@@ -5,13 +5,14 @@
  * @param {String} selector
  * @returns {Array}
  */
-export default (nodes: NodeList|HTMLCollection|Array, selector: string): Array<Element> => {
+export default (nodes: NodeList|HTMLCollection|Array<HTMLElement>, selector: string): Array<HTMLElement> => {
   if (!(nodes instanceof NodeList || nodes instanceof HTMLCollection || nodes instanceof Array)) {
     throw new Error('You must provide a nodeList/HTMLCollection/Array of elements to be filtered.')
   }
+  let elArray:Element[] = Array.from(nodes)
```
Comment (lukasoppermann) [inline src/filter.ts]: Sorry to be picky, but can you rename `elArray` to just `elements` I think it is more consistent and makes it easier to understand.

File: src/html5sortable.ts [hunk 21]
```diff
@@ -39,7 +39,7 @@ var _removeItemEvents = function (items) {
 }
 /**
  * Remove event handlers from sortable
- * @param {Element} sortable a single sortable
+ * @param {HTMLElement} sortable a single sortable
  */
 var _removeSortableEvents = function (sortable) {
```
Comment (lukasoppermann) [inline src/html5sortable.ts]: We could add a type here: `function (sortable: HTMLElement) {`

**Comment by jmuzsik** [comment]
Think I finally understand interactive rebase, that's mainly thanks to you, thanks! And I did the quick code fix on github by spontaneously, probably should have did it in my own editor.

File: src/eventListener.ts [hunk 22]
```diff
@@ -1,9 +1,10 @@
 /**
- * @param {Array|Element} element
- * @param {Array|string} event
+ * @param {Array|HTMLElement} element
+ * @param {string} event
  * @param {Function} callback
  */
-function addEventListener (element, event, callback) {
+
+function addEventListener (element: Array<HTMLElement>|HTMLElement, event:string, callback: () => void) {
```
Comment (lukasoppermann) [inline src/eventListener.ts]: Are you sure the callback part is correct? Can you explain it? Shouldn't it be callback: Function?

I guess you look this up somewhere so I probably just don't know about it, maybe you can point me to the resource?

File: src/serialize.ts [hunk 23]
```diff
@@ -19,9 +20,15 @@ export default (sortableContainer: Element, customItemSerializer: Function = (se
   }
   // get options
   let options = _data(sortableContainer, 'opts')
+
```
Comment (lukasoppermann) [inline src/serialize.ts:22]: This type guard is not nessesary, as the options part will be replaced with the store.

File: tsconfig.json [hunk 24]
```diff
@@ -1,5 +1,11 @@
 {
   "compilerOptions": {
-    "module": "commonjs"
-  }
+    "module": "commonjs",
+    "target": "ESNext",
+    "strictNullChecks": true
+  },
+  "typeRoots": [
+    "./src/interfaces",
```
Comment (lukasoppermann) [inline tsconfig.json]: This should be types now. 😉

**Review by lukasoppermann — COMMENTED** [review]
Just some minor last things and we can merge it, very nice work.

File: src/types/main.d.ts [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+    hoverClass: boolean,
+    debounce: number,
+    maxItems: number,
+    itemSerializer: void,
+    containerSerializer: void,
+    items: string
+}
+
+// Can these functions be better defined?
+interface events {
```
Comment (lukasoppermann) [inline src/types/main.d.ts]: I think this is not needed, as we can use `DragEvent` which should have those functions, right?

File: src/types/main.d.ts [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+ * typescriptlang.org/docs/handbook/declaration-merging.html
+ * ie. if this is done new properties can be 
+ * added to previously defined objects
+*/
+interface HTMLElement {
+    isSortable: boolean,
+    h5s: h5s
+}
+
+interface Node {
```
Comment (lukasoppermann) [inline src/types/main.d.ts]: Node does not have `querySelector`, so we need to remove this.

File: src/types/main.d.ts [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+interface h5s {
+    events: events,
+    data: data
+}
+/* 
+ * typescriptlang.org/docs/handbook/declaration-merging.html
+ * ie. if this is done new properties can be 
+ * added to previously defined objects
+*/
+interface HTMLElement {
```
Comment (lukasoppermann) [inline src/types/main.d.ts]: I think we should not extend `HTMLElement`. Rather, in the future we should have an interface for the sortable class #326

File: src/types/main.d.ts [hunk 28]
```diff
@@ -0,0 +1,59 @@
+/*
+ * For default interfaces that are frequently used and extensions:
+ *   - HTMLElement: h5s and its inner objects.
+ *   - Extension of objects that are external to the library.
+*/
+
+interface opts {
```
Comment (lukasoppermann) [inline src/types/main.d.ts]: could you rename this one to `configuration`?  I know we currently use otps but I would like to change this as well. Thanks

File: src/eventListener.ts (hunk 22 above)
Comment (jmuzsik) [reply src/eventListener.ts]: https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html

And it says to put parameters as well if you know what the callback will be, but it appears that the function is not always the same.

File: src/serialize.ts (hunk 23 above)
Comment (jmuzsik) [reply src/serialize.ts:22]: ok deleted

File: tsconfig.json (hunk 24 above)
Comment (jmuzsik) [reply tsconfig.json]: Maybe best to delete this b/c it does not appear to affect my linter.

File: src/types/main.d.ts (hunk 25 above)
Comment (jmuzsik) [reply src/types/main.d.ts]: ok deleted

File: src/types/main.d.ts (hunk 27 above)
Comment (jmuzsik) [reply src/types/main.d.ts]: I can look into that next

**Review by lukasoppermann — APPROVED** [review]

**Comment by lukasoppermann** [comment]
Looks very good, thank you for all the work @jMuzsik 👍

**Pull request merged**